### PR TITLE
Add completely new push syntax

### LINKED ISSUE #341: Add the ability to use a custom HTTP Client to the push
Currently, the push to gateway function uses http.DefaultClient to make the request to the pushgateway.
By abling to use a custom client, there can be more flexibility on how requests are sent.
For my specific usage when I encoutered the need, I am using the client from GAE, which doesn't allow at all using the http.DefaultClient, which makes the library unusable in GAE

### PR DESCRIPTION
**Pull request by beorn7**
This allows adding more options in elegant ways, showcased here by
HTTP basic auth and by injecting a custom http.Client.

Fixes #341 and #372.

@grobie @alitari @shlimp

### CODE REVIEW

**Comment by beorn7** [comment]
See #341 in particular to see the idea behind this.

File: prometheus/push/example_add_from_gatherer_test.go [hunk 1]
```diff
@@ -67,18 +71,16 @@ func ExampleAddFromGatherer() {
 	if err != nil {
 		fmt.Println("DB backup failed:", err)
 	} else {
-		// Only now register successTime.
-		registry.MustRegister(successTime)
+		// Add succesTime to pusher only in case of success.
```
Comment (brian-brazil) [inline prometheus/push/example_add_from_gatherer_test.go]: success

File: prometheus/push/example_add_from_gatherer_test.go [hunk 2]
```diff
@@ -67,18 +71,16 @@ func ExampleAddFromGatherer() {
...
+		// We could as well register it with the registry.
+		// This example, however, demonstrates that you can
+		// mix Gatherers and Collectors when handling a Pusher.
+		pusher.Collector(successTime)
```
Comment (brian-brazil) [inline prometheus/push/example_add_from_gatherer_test.go]: Its not clear to me at first glance exactly what this call is doing, is it pushing or is it registering?
`AddCollector` may be clearer

File: prometheus/push/push.go [hunk 3]
```diff
@@ -44,57 +52,164 @@ import (
...
+		if strings.Contains(value, "/") {
+			p.error = fmt.Errorf("value of grouping label %s contains '/': %s", name, value)
+			return p
 		}
+		p.grouping[name] = value
+	}
+	return p
+}
+
+// HostnameGrouping adds a label pair with “instance” as the name and the
```
Comment (brian-brazil) [inline prometheus/push/push.go]: The reason this existed was to allow for past mis-uses of the pgw when the pgw api changed quite some time ago. Now might be the time to remove it.

File: prometheus/push/example_add_from_gatherer_test.go (hunk 1 above)
Comment (beorn7) [reply prometheus/push/example_add_from_gatherer_test.go]: Fixed.

File: prometheus/push/example_add_from_gatherer_test.go (hunk 2 above)
Comment (beorn7) [reply prometheus/push/example_add_from_gatherer_test.go]: I had considered that name, too, but in the end I decided to go for the approach to use a verb only for the methods that are actually performing the push to the PGW (Push/Add) and use nouns for the  methods that configure the push. This also makes the methods shorter. Furthermore, using the verb "add" specifically will easily get confused with the `Add` method.

Hence, I do thing the current naming is the clearest and easiest to handle. Does somebody else have an opinion on this?

File: prometheus/push/push.go (hunk 3 above)
Comment (beorn7) [reply prometheus/push/push.go]: Personally, I don't feel strongly. I vaguely remember that some people did feel strongly about it. @grobie @juliusv I think you were part of the debate.

If nobody speaks up now, I'll happily remove it.

Comment (juliusv) [reply prometheus/push/push.go]: I don't remember anything, but I'm for removing it. As the author of https://prometheus.io/docs/practices/pushing/, I really want to discourage this use :)

Comment (beorn7) [reply prometheus/push/push.go]: OK, that's quorum. I have removed `HostnameGrouping`, adjusted the tests and examples, and edited the deprecation notice on the old `HostnameGroupingKey`.

**Pull request merged**